sycl/unittests/accessor/HostAccessorReverseIterator.cpp

```diff
@@ -0,0 +1,575 @@
+
+    sycl::id<3> shapeToCheck(
+        (Dimensions > 2 ? copyShape[Dimensions - 3] : 1) + offsetToUse[0],
+        (Dimensions > 1 ? copyShape[Dimensions - 2] : 1) + offsetToUse[1],
+        copyShape[Dimensions - 1] + offsetToUse[2]);
+
+    for (size_t z = offsetToUse[0]; z < shapeToCheck[0]; ++z) {
+      for (size_t y = offsetToUse[1]; y < shapeToCheck[1]; ++y) {
+        for (size_t x = offsetToUse[2]; x < shapeToCheck[2]; ++x) {
+          auto value = accessHelper<Dimensions>(fullAccessor, z, y, x);
+          ASSERT_EQ(copied[linearId], value);
+          ++linearId;
+        }
+      }
+    }
+  }
+
+  template <int TotalDimensions, int CurrentDimension = 3, typename Container,
+            typename... Indices>
+  auto &&accessHelper(Container &&C, int Idx, Indices... Ids) {
+    if constexpr (CurrentDimension > TotalDimensions) {
+      (void)Idx;
+      return accessHelper<TotalDimensions, CurrentDimension - 1>(C, Ids...);
+    } else
+      return accessHelper<TotalDimensions, CurrentDimension - 1>(C[Idx],
+                                                                 Ids...);
+  }
+
+  template <int TotalDimensions, int CurrentDimension = 3, typename Container>
+  auto &&accessHelper(Container &&C, int Idx) {
+    return C[Idx];
+  }
+};
+
+// FIXME: consider turning this into parameterized test to check various
+// accessor types
+TEST_F(HostAccessorReverseIteratorTest, IteratorTraits) {
+  using IteratorT = sycl::host_accessor<int>::reverse_iterator;
+  ASSERT_TRUE(
+      (std::is_same_v<sycl::host_accessor<int>::difference_type,
+                      std::iterator_traits<IteratorT>::difference_type>));
+  ASSERT_TRUE((std::is_same_v<sycl::host_accessor<int>::value_type,
+                              std::iterator_traits<IteratorT>::value_type>));
+  ASSERT_TRUE((std::is_same_v<sycl::host_accessor<int>::value_type *,
+                              std::iterator_traits<IteratorT>::pointer>));
+  ASSERT_TRUE((std::is_same_v<sycl::host_accessor<int>::reference,
+                              std::iterator_traits<IteratorT>::reference>));
+  ASSERT_TRUE(
+      (std::is_same_v<std::random_access_iterator_tag,
+                      std::iterator_traits<IteratorT>::iterator_category>));
+}
+
+// Based on requirements listed at
+// https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator
+TEST_F(HostAccessorReverseIteratorTest,
+       LegacyRandomAccessIteratorRequirements) {
+  using IteratorT = sycl::host_accessor<int>::reverse_iterator;
+  IteratorT It;
+  auto &RefToIt = It;
+  ASSERT_TRUE((std::is_same_v<IteratorT &, decltype(RefToIt += 3)>));
+  ASSERT_TRUE((std::is_same_v<IteratorT, decltype(It + 3)>));
+  ASSERT_TRUE((std::is_same_v<IteratorT, decltype(3 + It)>));
+  ASSERT_TRUE((std::is_same_v<IteratorT &, decltype(RefToIt -= 3)>));
+  ASSERT_TRUE((std::is_same_v<IteratorT, decltype(It - 3)>));
+  IteratorT It2;
+  ASSERT_TRUE((std::is_same_v<std::iterator_traits<IteratorT>::difference_type,
```

Comment:
This code doesn't compile on Windows.
Post-commit full log (https://github.com/intel/llvm/actions/runs/3941506279/jobs/6744005551)

```
2023-01-17T17:20:48.8570283Z D:\github\_work\llvm\llvm\src\sycl\unittests\accessor\HostAccessorReverseIterator.cpp(208): error C2678: binary '-': no operator found which takes a left-hand operand of type 'IteratorT' (or there is no acceptable conversion)
2023-01-17T17:20:48.8571772Z C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.34.31933\include\xutility(1318): note: could be 'std::reverse_iterator<sycl::_V1::detail::accessor_iterator<int,1>> std::reverse_iterator<sycl::_V1::detail::accessor_iterator<int,1>>::operator -(const __int64) noexcept(<expr>) const'
2023-01-17T17:20:48.8573712Z D:\github\_work\llvm\llvm\build\include\sycl/detail/accessor_iterator.hpp(108): note: or       'sycl::_V1::detail::accessor_iterator<int,1> sycl::_V1::detail::accessor_iterator<int,1>::operator -(const sycl::_V1::detail::accessor_iterator<int,1> &,sycl::_V1::detail::accessor_iterator<int,1>::difference_type)' [found using argument-dependent lookup]
2023-01-17T17:20:48.8575028Z C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.34.31933\include\xutility(1483): note: or       'unknown-type std::operator -(const std::reverse_iterator<_BidIt> &,const std::reverse_iterator<_BidIt2> &) noexcept(<expr>)'
2023-01-17T17:20:48.8576395Z D:\github\_work\llvm\llvm\src\sycl\unittests\accessor\HostAccessorReverseIterator.cpp(208): note: Failed to specialize function template 'unknown-type std::operator -(const std::reverse_iterator<_BidIt> &,const std::reverse_iterator<_BidIt2> &) noexcept(<expr>)'
2023-01-17T17:20:48.8577626Z D:\github\_work\llvm\llvm\src\sycl\unittests\accessor\HostAccessorReverseIterator.cpp(208): note: With the following template arguments:
2023-01-17T17:20:48.8578923Z D:\github\_work\llvm\llvm\src\sycl\unittests\accessor\HostAccessorReverseIterator.cpp(208): note: '_BidIt1=sycl::_V1::detail::accessor_iterator<int,1>'
2023-01-17T17:20:48.8579877Z D:\github\_work\llvm\llvm\src\sycl\unittests\accessor\HostAccessorReverseIterator.cpp(208): note: '_BidIt2=sycl::_V1::detail::accessor_iterator<int,1>'
2023-01-17T17:20:48.8581326Z C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.34.31933\include\xutility(3649): note: or       'unknown-type std::operator -(const std::move_iterator<_Iter> &,const std::move_iterator<_Iter2> &) noexcept(<expr>)'
2023-01-17T17:20:48.8583027Z D:\github\_work\llvm\llvm\src\sycl\unittests\accessor\HostAccessorReverseIterator.cpp(208): note: 'unknown-type std::operator -(const std::move_iterator<_Iter> &,const std::move_iterator<_Iter2> &) noexcept(<expr>)': could not deduce template argument for 'const std::move_iterator<_Iter> &' from 'IteratorT'
2023-01-17T17:20:48.8584287Z C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.34.31933\include\complex(1411): note: or       'std::complex<_Other> std::operator -(const std::complex<_Other> &,const std::complex<_Other> &)'
2023-01-17T17:20:48.8585865Z D:\github\_work\llvm\llvm\src\sycl\unittests\accessor\HostAccessorReverseIterator.cpp(208): note: 'std::complex<_Other> std::operator -(const std::complex<_Other> &,const std::complex<_Other> &)': could not deduce template argument for 'const std::complex<_Other> &' from 'IteratorT'
2023-01-17T17:20:48.8587350Z C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.34.31933\include\complex(1418): note: or       'std::complex<_Other> std::operator -(const std::complex<_Other> &,const _Ty &)'
2023-01-17T17:20:48.8588549Z D:\github\_work\llvm\llvm\src\sycl\unittests\accessor\HostAccessorReverseIterator.cpp(208): note: 'std::complex<_Other> std::operator -(const std::complex<_Other> &,const _Ty &)': could not deduce template argument for 'const std::complex<_Other> &' from 'IteratorT'
2023-01-17T17:20:48.8589993Z C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.34.31933\include\complex(1425): note: or       'std::complex<_Other> std::operator -(const _Ty &,const std::complex<_Other> &)'
2023-01-17T17:20:48.8591189Z D:\github\_work\llvm\llvm\src\sycl\unittests\accessor\HostAccessorReverseIterator.cpp(208): note: 'std::complex<_Other> std::operator -(const _Ty &,const std::complex<_Other> &)': could not deduce template argument for 'const std::complex<_Other> &' from 'IteratorT'
2023-01-17T17:20:48.8592647Z C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.34.31933\include\complex(1479): note: or       'std::complex<_Other> std::operator -(const std::complex<_Other> &)'
2023-01-17T17:20:48.8593673Z D:\github\_work\llvm\llvm\src\sycl\unittests\accessor\HostAccessorReverseIterator.cpp(208): note: while trying to match the argument list '(IteratorT, IteratorT)'
```
@maarquitos14, please, fix as soon as possible.